tests/providers/conftest.py

```diff
@@ -0,0 +1,104 @@
+import os
+import re
+from typing import Type, Tuple
+
+import pytest
+
+from exchange import Message
+from exchange.providers import Usage, Provider
+
+OPENAI_API_KEY = "test_openai_api_key"
+OPENAI_ORG_ID = "test_openai_org_key"
+OPENAI_PROJECT_ID = "test_openai_project_id"
+
+
+@pytest.fixture
+def default_openai_env(monkeypatch):
+    """
+    This fixture prevents OpenAIProvider.from_env() from erring on missing
+    environment variables.
+
+    When running VCR tests for the first time or after deleting a cassette
+    recording, set required environment variables, so that real requests don't
+    fail. Subsequent runs use the recorded data, so don't them.
+    """
+    if "OPENAI_API_KEY" not in os.environ:
+        monkeypatch.setenv("OPENAI_API_KEY", OPENAI_API_KEY)
+
+
+AZURE_ENDPOINT = "https://test.openai.azure.com"
+AZURE_DEPLOYMENT_NAME = "test-azure-deployment"
+AZURE_API_VERSION = "2024-05-01-preview"
+AZURE_API_KEY = "test_azure_api_key"
+
+
+@pytest.fixture
+def default_azure_env(monkeypatch):
+    """
+    This fixture prevents AzureProvider.from_env() from erring on missing
+    environment variables.
+
+    When running VCR tests for the first time or after deleting a cassette
+    recording, set required environment variables, so that real requests don't
+    fail. Subsequent runs use the recorded data, so don't them.
+    """
+    if "AZURE_CHAT_COMPLETIONS_HOST_NAME" not in os.environ:
+        monkeypatch.setenv("AZURE_CHAT_COMPLETIONS_HOST_NAME", AZURE_ENDPOINT)
+    if "AZURE_CHAT_COMPLETIONS_DEPLOYMENT_NAME" not in os.environ:
+        monkeypatch.setenv("AZURE_CHAT_COMPLETIONS_DEPLOYMENT_NAME", AZURE_DEPLOYMENT_NAME)
+    if "AZURE_CHAT_COMPLETIONS_DEPLOYMENT_API_VERSION" not in os.environ:
+        monkeypatch.setenv("AZURE_CHAT_COMPLETIONS_DEPLOYMENT_API_VERSION", AZURE_API_VERSION)
+    if "AZURE_CHAT_COMPLETIONS_KEY" not in os.environ:
+        monkeypatch.setenv("AZURE_CHAT_COMPLETIONS_KEY", AZURE_API_KEY)
+
+
+@pytest.fixture(scope="module")
+def vcr_config():
+    """
+    This scrubs sensitive data and gunzips bodies when in recording mode.
+
+    Without this, you would leak cookies and auth tokens in the cassettes.
+    Also, depending on the request, some responses would be binary encoded
+    while others plain json. This ensures all bodies are human-readable.
+    """
+    return {
+        "decode_compressed_response": True,
+        "filter_headers": [
+            ("authorization", "Bearer " + OPENAI_API_KEY),
+            ("openai-organization", OPENAI_ORG_ID),
+            ("openai-project", OPENAI_PROJECT_ID),
+            ("cookie", None),
+        ],
+        "before_record_request": scrub_request_url,
+        "before_record_response": scrub_response_headers,
+    }
+
+
+def scrub_request_url(request):
+    """
+    This scrubs sensitive request data in provider-specific way. Note that headers
+    are case-sensitive!
+    """
+    if "openai" in request.uri:
```

Comment:
Using the presence of `openai` in the URI to indicate azure is still blowing my mind, especially so since it's not obvious the uri doesn't contain the host which could otherwise have been `api.openai.com`.

Perhaps `request.headers["host"] == os.environ.get("AZURE_CHAT_COMPLETIONS_HOST_NAME")` is an option? Admittedly not great either, just a note to see if there isn't anything to improve here

Reply:
this was a goof on my side, I meant to put openai.azure.com which I think we can keep as a condition . The problem with using that ENV variable is that it is weird, like it isn't a hostname, rather a base URL.